lib/puppet/provider/cisco_ospf_area/cisco.rb

```diff
@@ -0,0 +1,168 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'cisco_node_utils' if Puppet.features.cisco_node_utils?
+begin
+  require 'puppet_x/cisco/autogen'
+rescue LoadError # seen on master, not on agent
+  # See longstanding Puppet issues #4248, #7316, #14073, #14149, etc. Ugh.
+  require File.expand_path(File.join(File.dirname(__FILE__), '..', '..', '..',
+                                     'puppet_x', 'cisco', 'autogen.rb'))
+end
+
+Puppet::Type.type(:cisco_ospf_area).provide(:cisco) do
+  desc 'The Cisco OSPF area provider.'
+
+  confine feature: :cisco_node_utils
+  defaultfor operatingsystem: :nexus
+
+  mk_resource_methods
+
+  OSPF_AREA_NON_BOOL_PROPS = [
+    :authentication,
+    :default_cost,
+    :filter_list_in,
+    :filter_list_out,
+  ]
+  OSPF_AREA_BOOL_PROPS = [
+    :stub,
+    :stub_no_summary,
+  ]
+  OSPF_AREA_ARRAY_FLAT_PROPS = [
+    :range
+  ]
+
+  OSPF_AREA_ALL_PROPS = OSPF_AREA_NON_BOOL_PROPS + OSPF_AREA_BOOL_PROPS +
+                        OSPF_AREA_ARRAY_FLAT_PROPS
+
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:non_bool, self, '@nu',
+                                            OSPF_AREA_NON_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:bool, self, '@nu',
+                                            OSPF_AREA_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:array_flat, self, '@nu',
+                                            OSPF_AREA_ARRAY_FLAT_PROPS)
+
+  def initialize(value={})
+    super(value)
+    ospf = @property_hash[:ospf]
+    vrf = @property_hash[:vrf]
+    area = @property_hash[:area]
+    @nu = Cisco::RouterOspfArea.areas[ospf][vrf][area] unless
+      ospf.nil? || vrf.nil? || area.nil?
+    @property_flush = {}
+  end
+
+  def self.properties_get(ospf, vrf, area, nu_obj)
+    debug "Checking ospf instance, #{ospf} #{vrf} #{area}"
+    current_state = {
+      name:   "#{ospf} #{vrf} #{area}",
+      ospf:   ospf,
+      vrf:    vrf,
+      area:   area,
+      ensure: :present,
+    }
+    # Call node_utils getter for each property
+    OSPF_AREA_NON_BOOL_PROPS.each do |prop|
+      current_state[prop] = nu_obj.send(prop)
+    end
+    OSPF_AREA_BOOL_PROPS.each do |prop|
+      val = nu_obj.send(prop)
+      if val.nil?
+        current_state[prop] = nil
+      else
+        current_state[prop] = val ? :true : :false
+      end
+    end
+    OSPF_AREA_ARRAY_FLAT_PROPS.each do |prop|
+      current_state[prop] = nu_obj.send(prop)
+    end
+    # nested array properties
+    current_state[:range] = nu_obj.range
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    area_instances = []
+    Cisco::RouterOspfArea.areas.each do |ospf, vrf|
+      vrf.each do |name, areas|
+        areas.each do |area|
+          area_instances << properties_get(ospf, name, area[0], area[1])
```

Comment:
This code block really confuses me. 'name' is rarely a helpful variable name when you have several objects to deal with. Consider this instead:

```
   Cisco::RouterOspfArea.areas.each do |ospf, vrfs|
     vrfs.each do |vrf, areas|
     areas.each do |area, nu_obj|
       area_instances << properties_get(ospf, vrf, area, nu_obj)
```